### ISSUE TITLE
[sdk] Bug when trying to iterate a list of dictionaries with ParallelFor

### ISSUE THREAD

**Issue by lucasvbalves** [kubeflow/pipelines#9366]
### Environment

*  KFP SDK version: 
kfp==2.0.0b16
*  All dependencies version:
kfp==2.0.0b16
kfp-pipeline-spec==0.2.2
kfp-server-api==2.0.0b1

### Steps to reproduce

When running the code snippet below the following error is raised:

`kfp.components.types.type_utils.InconsistentTypeException: Incompatible argument passed to the input 'val_a' of component 'add': Argument type 'STRING' is incompatible with the input type 'NUMBER_INTEGER'`

```python
@dsl.component()
def add(val_a: int, val_b: int) -> int:
    return val_a + val_b


@dsl.pipeline()
def model_training_pipeline() -> None:
    with dsl.ParallelFor(
        items=[{"a": 1, "b": 10}, {"a": 2, "b": 20}], parallelism=1
    ) as item:
        task = add(val_a=item.a, val_b=item.b)


compiler.Compiler().compile(
    pipeline_func=model_training_pipeline, package_path="/app/pipeline.yaml"
)
```

### Expected result

According to the [ParallelFor documentation](https://kubeflow-pipelines.readthedocs.io/en/2.0.0b16/source/dsl.html#kfp.dsl.ParallelFor), the code sample above should compile without errors. The `add` component should receive the values of the dictionaries as integer arguments.

### Materials and Reference

The code snippet below is a modification of the code snippet above, changing the `add` component to accept string arguments. 

```python
@dsl.component()
def add(val_a: str, val_b: str) -> int:
    return int(val_a) + int(val_b)


@dsl.pipeline()
def model_training_pipeline() -> None:
    with dsl.ParallelFor(
        items=[{"a": 1, "b": 10}, {"a": 2, "b": 20}], parallelism=1
    ) as item:
        task = add(val_a=item.a, val_b=item.b)


compiler.Compiler().compile(
    pipeline_func=model_training_pipeline, package_path="/app/pipeline.yaml"
)
```

The pipeline compiles without errors with this modification, however it fails to run in Google Vertex Pipelines. The `add` component doesn't even run and throws the following error in the UI:

`Failed to evaluate the expression with error: INVALID_ARGUMENT: Failed to parseJson from string.; Failed to evaluate the parameter_expression_selector.`

As the component's code is not even executed, it seems that the problem occurs when executing the DAG.

Here is the content of the `pipeline.yaml` that was compiled.

```yaml
# PIPELINE DEFINITION
# Name: model-training-pipeline
components:
  comp-add:
    executorLabel: exec-add
    inputDefinitions:
      parameters:
        val_a:
          parameterType: STRING
        val_b:
          parameterType: STRING
    outputDefinitions:
      parameters:
        Output:
          parameterType: NUMBER_INTEGER
  comp-for-loop-2:
    dag:
      tasks:
        add:
          cachingOptions:
            enableCache: true
          componentRef:
            name: comp-add
          inputs:
            parameters:
              val_a:
                componentInputParameter: pipelinechannel--loop-item-param-1
                parameterExpressionSelector: parseJson(string_value)["a"]
              val_b:
                componentInputParameter: pipelinechannel--loop-item-param-1
                parameterExpressionSelector: parseJson(string_value)["b"]
          taskInfo:
            name: add
    inputDefinitions:
      parameters:
        pipelinechannel--loop-item-param-1:
          parameterType: STRUCT
deploymentSpec:
  executors:
    exec-add:
      container:
        args:
        - --executor_input
        - '{{$}}'
        - --function_to_execute
        - add
        command:
        - sh
        - -c
        - "\nif ! [ -x \"$(command -v pip)\" ]; then\n    python3 -m ensurepip ||\
          \ python3 -m ensurepip --user || apt-get install python3-pip\nfi\n\nPIP_DISABLE_PIP_VERSION_CHECK=1\
          \ python3 -m pip install --quiet     --no-warn-script-location 'kfp==2.0.0-beta.16'\
          \ && \"$0\" \"$@\"\n"
        - sh
        - -ec
        - 'program_path=$(mktemp -d)

          printf "%s" "$0" > "$program_path/ephemeral_component.py"

          python3 -m kfp.components.executor_main                         --component_module_path                         "$program_path/ephemeral_component.py"                         "$@"

          '
        - "\nimport kfp\nfrom kfp import dsl\nfrom kfp.dsl import *\nfrom typing import\
          \ *\n\ndef add(val_a: str, val_b: str) -> int:\n    return int(val_a) +\
          \ int(val_b)\n\n"
        image: python:3.7
pipelineInfo:
  name: model-training-pipeline
root:
  dag:
    tasks:
      for-loop-2:
        componentRef:
          name: comp-for-loop-2
        iteratorPolicy:
          parallelismLimit: 1
        parameterIterator:
          itemInput: pipelinechannel--loop-item-param-1
          items:
            raw: '[{"a": 1, "b": 10}, {"a": 2, "b": 20}]'
        taskInfo:
          name: for-loop-2
schemaVersion: 2.1.0
sdkVersion: kfp-2.0.0-beta.16

```

---

Impacted by this bug? Give it a 👍.

**Comment by msmendoza** [comment]
I solved this by using `import json` and then

```
with dsl.ParallelFor(
        items=[json.dump({"a": 1, "b": 10}), json.dump({"a": 2, "b": 20})], parallelism=1
    ) as item:
        task = add(val_a=item.a, val_b=item.b)
```

Hope it helps,

**Comment by hjerpe** [comment]
@msmendoza do you mean `json.dumps`? Otherwise there is an error thrown, TypeError: dump() missing 1 required positional argument: 'fp'

**Comment by alvarodemig** [comment]
hey @connor-mccarthy! any update on this?

**Comment by connor-mccarthy** [comment]
No update at this moment. I can prioritize a review of a fix PR from any contributor.

**Comment by stijntratsaertit** [comment]
Hi @connor-mccarthy, do you have any hints on what the underlying cause could be? Perhaps regarding `parameterExpressionSelector: parseJson(string_value)["a"]``not being able to accept an integer?

I've been investigating and looking for a fix, but I'm but bit stuck

**Comment by pokutuna** [comment]
I'd like to provide some information and a workaround for this issue.
It seems that passing a `list[dict]` with mixed data types (like `str` and `int`) to items is still not possible.

- Working:
  - `dsl.ParallelFor(items=[{"a": 1, "b": 2}, {"a": 3, "b": 4}], ...)`
  - `dsl.ParallelFor(items=[{"a": 1, "b": 2}, {"a": 3, "b": 4}], ...)`
- Not Working:
  - `dsl.ParallelFor(items=[{"a": 1, "b": "2"}, {"a": 3, "b": "4"}], ...)`

This was addressed in #10025 but wasn't merged.


As a workaround, we can use pipeline parameters.

This usually works because in real-world scenarios, we often want to run tasks in parallel based on some predefined data rather than writing values directly in ParallelFor.

Here's an example that works with Vertex AI Pipelines:

```python
import kfp
from kfp import dsl
from google.cloud import aiplatform

@dsl.component(base_image="python:3.11")
def print_params(index: int, name: str) -> None:
    print([index, name])

@dsl.pipeline()
def pipeline(tasks: list) -> None:
    with dsl.ParallelFor(items=tasks) as item:
        print_params(index=item.index, name=item.name_)

if __name__ == "__main__":
    aiplatform.init(project="...", location="...")

    dest = __file__.replace(".py", ".yaml")
    compiler = kfp.compiler.Compiler()
    compiler.compile(pipeline_func=pipeline, package_path=dest)

    tasks = [
        {"index":0, "name_":"task1"},
        {"index":1, "name_":"task2"},
    ]
    job = aiplatform.PipelineJob(
        template_path=dest,
        display_name='parallel-values-from-pipeline',
        parameter_values={"tasks": tasks},
    )
    job.submit()
```

Note: some names like `item.name` won't work for referencing.
That's why in the example, I use `item.name_` with an underscore.

The loop variable in `ParallelFor` (`item` in the example) is a `LoopParameterArgument`.
References like `item.index` or `item.name_` are resolved by the `__getattr__` implementation.
https://github.com/kubeflow/pipelines/blob/9cb5913d4757132e2c6e5f1c5fa2cdb7ad541f97/sdk/python/kfp/dsl/for_loop.py#L169

`LoopParameterArgument` inherits from classes `PipelineParameterChannel` and others, so the instance fields might be found before the given dictionary keys from the loop target.
You should avoid names like `name`, `task_name`, `value` and something others.
